src/libcadet/model/parts/TractorConvectionDispersionOperator.cpp

```diff
@@ -0,0 +1,1526 @@
+					const ParamType exchange_orig_dest_comp = static_cast<ParamType>(_exchangeMatrix[rad_orig * _nRad * _nComp + rad_dest * _nComp + comp]);
+					if (cadet_likely(exchange_orig_dest_comp > 0.0))
+					{
+						*resCur_orig += exchange_orig_dest_comp * yCur_orig[0];
+						*resCur_dest -= exchange_orig_dest_comp * yCur_orig[0];
+
+						// TODO: @Sam, please check this
+						if (wantJac)
+						{
+							_jacC.centered(offsetCur_orig, 0) += static_cast<double>(exchange_orig_dest_comp);
+							_jacC.centered(offsetCur_orig, offsetCur_dest) -= static_cast<double>(exchange_orig_dest_comp);
+							_jacC.centered(offsetCur_dest, 0) -= static_cast<double>(exchange_orig_dest_comp);
+							_jacC.centered(offsetCur_dest, offsetCur_orig) += static_cast<double>(exchange_orig_dest_comp);
+						}
+					}
+                }
+
+            }
+
+		}
+	}
+
+	return 0;
+}
+
+template <typename ParamType>
+void TractorConvectionDispersionOperator::setSparsityPattern()
+{
+	// Note that we have to increase the lower non-zeros by 1 because the WENO stencil is applied to the
+	// right cell face (lower + 1 + upper) and to the left cell face (shift the stencil by -1 because influx of cell i
+	// is outflux of cell i-1)
+	// We also have to make sure that there's at least one sub and super diagonal for the dispersion term
+	const unsigned int lowerNonZeros = std::max(_weno.lowerBandwidth() + 1u, 1u);
+	const unsigned int upperNonZeros = std::max(_weno.upperBandwidth(), 1u);
+	// Total number of non-zeros per row is WENO stencil (lowerNonZeros + 1u + upperNonZeros) + radial dispersion (2)
+	cadet::linalg::SparsityPattern pattern(_nComp * _nCol * _nRad, lowerNonZeros + 1u + upperNonZeros + 2u);
+
+	// Handle convection, axial dispersion (WENO)
+	for (unsigned int i = 0; i < _nRad; ++i)
+		cadet::model::parts::convdisp::sparsityPattern(pattern.row(i * _nComp), _nComp, _nCol, _nComp * _nRad, static_cast<double>(_curVelocity[i]), _weno);
+
+	// Handle radial dispersion
+	// TODO: Restore original setup for radial dispersion and add separate block for Tractor
+	if (_nRad > 1)
+	{
+		for (unsigned int col = 0; col < _nCol; ++col)
+		{
+			// Axial column element
+			const unsigned int idxColBlock = col * _nRad * _nComp;
+
+			// Connecting from all compartments (orig) to all compartments (dest)
+			for (unsigned int rad_orig = 0; rad_orig < _nRad ; ++rad_orig)
+			{
+				const unsigned int idxColRadBlock_orig = idxColBlock + rad_orig * _nComp;
+
+				for (unsigned int rad_dest = 0; rad_dest < _nRad ; ++rad_dest)
+				{
+					const unsigned int idxColRadBlock_dest = idxColBlock + rad_dest * _nComp;
+
+					// Don't connect compartments with themselves
+					if (idxColRadBlock_dest == idxColRadBlock_orig)
+						continue;
+
+					for (unsigned int comp = 0; comp < _nComp; ++comp)
+					{
+						const ParamType exchange_orig_dest_comp = static_cast<ParamType>(_exchangeMatrix[rad_orig * _nRad * _nComp + rad_dest * _nComp + comp]);
```

Comment:
This will always be `active`. In fact, it will crash or produce garbage if `ParamType` is anything other than `active`. So no need for this function to be a template.